pkg/controller/miganalytic/restic_executor.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2021 Red Hat Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package miganalytic
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/konveyor/mig-controller/pkg/compat"
+	"github.com/konveyor/mig-controller/pkg/pods"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ResticPodLabelKey is the key of the label used to discover Restic pod
+	ResticPodLabelKey = "name"
+	// ResticPodLabelValue is the value of the label used to discover Restic pod
+	ResticPodLabelValue = "restic"
+)
+
+// ResticDFCommandExecutor uses Restic pods to run DF command
+type ResticDFCommandExecutor struct {
+	// Namespace is the ns in which Restic pods are present
+	Namespace string
+	// Client to interact with Restic pods
+	Client compat.Client
+	// ResticPodReferences is a local cache of known Restic pods
+	ResticPodReferences map[string]*corev1.Pod
+}
+
+// DF given a podRef and a list of volumes, runs df command, returns with structured command context
+// any errors running the df command are suppressed here. DFCommand.stdErr field should be used to determine failure
+func (r *ResticDFCommandExecutor) DF(podRef *corev1.Pod, persistentVolumes []MigAnalyticPersistentVolumeDetails) DFCommand {
+	// TODO: use the appropriate block size based on PVCs
+	dfCmd := DFCommand{
+		BaseLocation: "/host_pods",
+		BlockSize:    DecimalSIMega,
+		StdOut:       "",
+		StdErr:       "",
+	}
+	cmdString := dfCmd.PrepareDFCommand(persistentVolumes)
+	restCfg := r.Client.RestConfig()
+	podCommand := pods.PodCommand{
+		Pod:     podRef,
+		RestCfg: restCfg,
+		Args:    cmdString,
+	}
+	err := podCommand.Run()
+	if err != nil {
+		log.Info(
```

Comment:
how does the user know that we are unsuccessful in determining the current pod volumes, and why

Can't this also error for other reasons like pod does not exist or unable to connect to pod or some other reason?

Reply:
@shawn-hurley according to current implementation, if this happens, all PVCs on this node will be added to Warning indicating that we failed to gather data for these volumes. It's either we succeed or we fail. Maybe we can log the message here, but unsure how to bubble up the detailed failure information all the way up until MigPlan.